### DISCUSSION TITLE
2022 GSOC Project Ideas

### DISCUSSION THREAD

**Discussion by kks32** [cb-geo/mpm#724] [Ideas]
# 2022 Google Summer of Code Project Ideas

##  eXtended Reality (XR) simulations
Extended Reality simulations involve augmenting real-world systems with simulations/computational tools. This project involves extracting real-world slope topography through image capture and running simulations in Material Point Method using the extracted topography. The results of which can be conveyed as runout projected onto the real-world slopes using Augmented Reality (AR). 

| **Intensity**        | **Type**                      | **Priority**              | **Involves**  | **Mentors**              |
| -------------        | -------------                         | ------------              | ------------- | -----------              |
| Moderate |  Long (350 hours)  |  High  | eXtended Reality simulations CB-Geo MPM. Developing XR framework for natural hazards and two-phase Material Point Method.       | [Krishna Kumar](https://github.com/kks32) and [Paul Navratil](https://github.com/pnav) | 

### Expected outcomes
- A prototype that integrates XR platform with MPM simulations to identify potential slopes that are unsafe. We'll augment MPM simulations based on real-world topography captured by the XR tool (Google VR headsets) to identify the slope failures. 

### Skills required
- Have a working knowledge of C++ and Python.
- Experience with tools such as Unity.
- Experience with rendering or visualization tools (ParaView, Houdini, Blender) is encouraged but not required.
- Experience with XR platforms are encouraged but not required.

##  In-situ visualization using Galaxy in MPM
In-situ visualization is a ray-tracing technique for visualizing the simulation data in real-time without requiring additional storage resources. This project will integrate distributed asynchronous ray-tracing with [TACC Galaxy](https://github.com/tacc/galaxy) into [CB-Geo MPM code](https://github.com/cb-geo/mpm) for rendering petascale simulations.

| **Intensity**           | **Type**                         | **Priority**              | **Involves**  | **Mentors**              |
| -------------            | -------------                   | ------------              | ------------- | -----------              |
| Moderate  | Medium  (175 hours)|  High  | Integrating Galaxy with existing rendering systems in CB-Geo MPM. Developing rendering of natural hazards such as landslides with in-situ viz interface.       | [Krishna Kumar](https://github.com/kks32) and [Paul Navratil](https://github.com/pnav) | 

### Expected outcomes
- A prototype that can do in-situ visualization of 100 million material points integrating MPM with TACC Galaxy. The prototype will also enable global illumination to improve the visual quality of rendering. 

### Skills required
- Have a working knowledge of C++ and Python.
- Experience with tools such as Unity.
- Experience with rendering or visualization tools (ParaView, Houdini, Blender) is encouraged but not required.
- Experience with distributed programming models (MPI).

# Benefits of working on CB-Geo MPM GSOC projects
Students who work on this project can expect their skillset to grow in
- Distributed asynchronous ray-tracing
- Petascale simulations
- In-situ visualization
- eXtended Reality
- Multiphase formulation in MPM

### Motivation
Current techniques for visualization of petascale simulations involve the post-hoc rendering of a temporal slice of a subset of the data from disk, which leads to a significant portion of information being disregarded and potentially lost. A large-scale simulation generates several terabytes of data that push commercial rendering engines like Blender, Mantra, and Arnold to the limit. The amount of data would be several orders of magnitude higher at petascale and exascale levels, making the simulation spend most of the supercomputing time doing I/O. 

### Technical Details
In-situ visualization is a rendering technique for visualizing the simulation data in real-time without requiring additional storage resources. Running the visualization and simulation in tandem avoids the bottleneck of data transfer. Furthermore, this approach allows for monitoring and interaction while the simulation is running, enabling scientists to modify simulation parameters and explore the effect on the phenomena in real-time. Ray tracing engines such as TACC Galaxy offer distributed asynchronous in-situ visualization capabilities for petascale simulations.

This project will integrate the in-situ data visualization and query framework for the HPC-scalable CB-Geo MPM code using the TACC Galaxy and Intel OSPRay libraries. Galaxy is a fully asynchronous distributed parallel rendering engine geared towards using full global illumination for large-scale visualization. Galaxy provides a performant distributed rendering using an actor model to render scenes across multiple MPI tasks asynchronously. Galaxy employs asynchronous frame buffer updates and a novel subtractive lighting model to achieve acceptable image quality even from the first ray generation, and the quality of rendering is continuously improved throughout the render epoch. This technology allows for transparent data communication across the network using a client/server model using sockets or MPI layers. The project involves effective communication of distributed data between the simulator and in-situ viz engine. Advanced sampling techniques with visual culling techniques are essential to render photorealistic MPM simulations at petascale. An a priori adaptive sampling method based on multiple viewpoints with visual culling of the rendered scenes will be tested with Monte Carlo sampling to render a billion particle simulation. 

The main goals of this project are to:
- Integrate in-situ visualization with TACC Galaxy in CB-Geo MPM.
- Perform an asynchronous distributed ray-tracing of billions of particles in a petascale simulation of natural hazards, such as the Oso landslide in Washington.
- Support context and user-specific visualization, i.e., different users may be interested in different aspects of the data sets.
- Develop modular and adaptive visualization infrastructure to automatically choose appropriate rendering configurations (lighting, camera position, material, textures, and identify regions of interest) for petascale problems.

### Benefits to project/community
The CB-Geo MPM currently supports VTK outputs and rendering support through Disney’s Partio library, which requires commercial rendering tools such as Houdini / Pixar Renderman. Enabling In-situ visualization capabilities in the CB-Geo Material Point Method code will open a new avenue of possibilities for running realistic petascale simulations, which has never been attempted. In-situ visualization with simultaneous multiple viewports will offer the ability to support context and user-specific visualization of the simulation results. Lessons learned from rendering at petascale with asynchronous communication and ray-tracing will benefit the wider community and significantly influence policymakers in devising strategies in the event of a natural hazard. The TACC Galaxy and Intel OSPray in-situ visualization libraries will enable a 100% open-source pipeline from input to final rendering.

### First steps
- Read our paper describing the process: [https://arxiv.org/abs/2109.02754](https://arxiv.org/abs/2109.02754), [Semi Implicit Scheme in MPM](https://jzhao.people.ust.hk/home/PDFs/2021-NAG-3207.pdf) and [https://arxiv.org/ftp/arxiv/papers/1909/1909.13380.pdf](https://arxiv.org/ftp/arxiv/papers/1909/1909.13380.pdf)
- Become familiar with existing ray tracing engines and various standard file formats for storage and visualization, such as VTP, VTK, BGEO, and HDF5.
- Become familiar with the [Intel OneAPI Rendering Toolkit (Embree, OSPRay, OpenVKL, etc)](https://software.intel.com/content/www/us/en/develop/tools/oneapi/rendering-toolkit.html) just to familiarize yourself with the entire stack and to know that those elements are out there.
- Become familiar with in-situ visualization libraries [TACC Galaxy](https://github.com/tacc/galaxy) and [Intel OSPRay](https://www.ospray.org/)
- Become familiar with current simulation and visualization infrastructure in CB-Geo MPM.
- Become familiar with running petascale simulations on TACC supercomputer

Please join our project discussions:

**Comment by Akhil-Sharma30** [comment]
Hey there @kks32 , I would like to work on some beginner-level issues to get an understanding of the codebase for the idea `In-situ visualization using Galaxy in MPM`. 
Could someone help me with this?

I am familiar with Unity Software and know C++ and Python programming Languages. 
Also, Have experience with rendering or visualization tools like Blender.

**Reply by kks32** [reply]
Hi @Akhil-Sharma30 We can start with writing `.obj` outputs. At the moment, we write `vtk` and `bgeo` files for rendering. If we use Blender, then we need `.obj` to render. A good exercise would be to use `vtkObjWriter` so you get familiar with the code and also use your blender skills to render animation?

**Reply by kks32** [reply]
you can join GSOC project discussion here.

**Comment by mfontaine218** [comment]
Hello! My name is Madeline Fontaine and I am a PhD Student at the University of Nevada, Reno. I study numerical methods for geomaterials, with a focus on geohazards such as debris flows. I have been using CB-Geo MPM for the past year and would like to contribute to the two-phase implementation as it would help me to better capture debris flow mechanics. I am interested in how visualization of large scale landslides could be improved and would like the opportunity to participate in the in-situ visualization updates proposed for GSoC 2022.
